flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionAttemptID.java

```diff
@@ -18,33 +18,77 @@
 
 package org.apache.flink.runtime.executiongraph;
 
-import org.apache.flink.util.AbstractID;
+import org.apache.flink.annotation.VisibleForTesting;
+import org.apache.flink.runtime.scheduler.strategy.ExecutionVertexID;
 
 import org.apache.flink.shaded.netty4.io.netty.buffer.ByteBuf;
+import org.apache.flink.util.Preconditions;
 
 /**
  * Unique identifier for the attempt to execute a tasks. Multiple attempts happen
  * in cases of failures and recovery.
  */
-public class ExecutionAttemptID extends AbstractID {
+public class ExecutionAttemptID implements java.io.Serializable {
 
 	private static final long serialVersionUID = -1169683445778281344L;
 
+	private final ExecutionVertexID executionVertexID;
+	private final int attemptNumber;
+
+	/**
+	 * Get a random execution attempt id.
+	 */
 	public ExecutionAttemptID() {
```

Comment:
This constructor is for tests only. Could we introduce a test util method `ExecutionGraphTestUtils#createRandomExecutionAttemptId()`? The returned id consists of a random `ExecutionVertexId` as well as a random `attemptNumber`. I think it better suites the tests which wants a random `ExecutionAttemptID`. And we can also avoid to add a test-only code path in production.

Reply:
Yes, that's a good point. However, it is currently used by production code and a lot of testing code paths. I would move this issue out of the scope of this PR. WDYT?

Reply:
Makes sense.